File: src/Tree/BinaryTreeBFSGuide/BinaryTreeLevelOrderTraversal/BinaryTreeLevelOrderTraversal.java
class Solution {
    public List<List<Integer>> levelOrder(TreeNode root) {
        List<List<Integer>> res = new LinkedList<>();
        //Base case
        if(root == null) return res;
        
        Queue<TreeNode> queue = new LinkedList<>();
        
        queue.add(root);
        
        while(!queue.isEmpty()){
            //Traverse current level ONLY
            int size = queue.size();
            List<Integer> level = new LinkedList<>();
            for(int i = 0; i < size; i++){
                TreeNode top = queue.poll();
                level.add(top.val);
                if(top.left != null){
                    queue.add(top.left);
                }
                if(top.right != null){
                    queue.add(top.right);
                }
            }
            //Add current level list onto the list
            res.add(level);
        }
        return res;   
    }
}

File: src/Tree/BinaryTreeBFSGuide/BinaryTreeLevelOrderTraversal/BinaryTreeLevelOrderTraversal.py
from collections import deque

class Solution:
    def levelOrder(self, root: Optional[TreeNode]) -> List[List[int]]:
        res = []
        
        if root is None:
            return res
        
        queue = deque([root])
        
        while len(queue) != 0:
            
            level_size = len(queue)
            
            level_list = []
            
            for idx in range(level_size):
                cur_node = queue.popleft()
                
                level_list.append(cur_node.val)
                
                if cur_node.left is not None:
                    queue.append(cur_node.left)
                    
                if cur_node.right is not None:
                    queue.append(cur_node.right)
                
            res.append(level_list)
                
        return res
                
